extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/KinesisRecordSupplier.java

```diff
@@ -847,36 +881,46 @@ private void checkIfClosed()
   }
 
   /**
-   * Returns an array with the content between the position and limit of "buffer". This may be the buffer's backing
-   * array itself. Does not modify position or limit of the buffer.
+   * This method must be called before a seek operation ({@link #seek}, {@link #seekToLatest}, or
+   * {@link #seekToEarliest}).
+   *
+   * When called, it will nuke the {@link #scheduledExec} that is shared by all {@link PartitionResource}, filters
+   * records from the buffer for partitions which will have a seek operation performed, and stops background fetch for
+   * each {@link PartitionResource} to prepare for the seek. If background fetch is not currently running, the
+   * {@link #scheduledExec} will not be re-created.
    */
-  private static byte[] toByteArray(final ByteBuffer buffer)
+  private void filterBufferAndResetBackgroundFetch(Set<StreamPartition<String>> partitions) throws InterruptedException
   {
-    if (buffer.hasArray()
-        && buffer.arrayOffset() == 0
-        && buffer.position() == 0
-        && buffer.array().length == buffer.limit()) {
-      return buffer.array();
-    } else {
-      final byte[] retVal = new byte[buffer.remaining()];
-      buffer.duplicate().get(retVal);
-      return retVal;
-    }
-  }
+    checkIfClosed();
```

Comment:
Added check for `backgroundFetchEnabled` to the if statement, but `checkIfClose` will still be called, is that cool? (since you put the comment on this line)

Reply:
Yeah, seems it should check to avoid unnecessary seek.